Trim whitespace characters from both ends of string

Amrysliu opened the pull request:
See: https://progress.opensuse.org/issues/51932

kalikiana commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -163,13 +163,20 @@ sub create {
+        }
+        my $value = $self->param($par);
+        $value =~ s/^\s+|\s+$//g;
+        $entry{$par} = $value;
```
I think it'd be a good idea to move getting the entries, settings and validation here to a separate method shared with the `update` sub because they do the exact same thing.

Amrysliu replied:
Maybe I should trim white space in this PR, after all according to Coolo's comments, I need to trim  white space in openqa.ini and workers.ini. Then I could create a new PR to improve the code about the repeated operation. Do you think it is ok?

kalikiana replied:
I'm just suggesting to move it in a sub, since you're making the exact same fix twice.

Although as long as you're adding the appropriate unit tests I don't mind if you do it in two PRs, your call.

Amrysliu replied:
I fix it in this PR. I think maybe it is convenient for merging code.

foursixnine commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -163,13 +163,20 @@ sub create {
+            next;
+        }
+        my $value = $self->param($par);
+        $value =~ s/^\s+|\s+$//g;
```
Just for the record, there's a trim function in Mojo::Util.

coolo commented:
Please check that openqa.ini and workers.ini get the same triming treatmeant. this should be fair unit tests.

kalikiana commented on t/api/05-machines.t:
```diff
@@ -163,6 +163,69 @@ is_deeply(
 $res = $t->delete_ok("/api/v1/machines/$machine_id")->status_is(404);    #not found
 
+subtest 'add and update machine with white spaces' => sub {
+    $res = $t->post_ok(
```
Remember you don't need the `$get` or `$res` variables here. `$t->tx` always has the result of the transaction.

Amrysliu replied:
> I think you need to rebase after we have updated the code style in another PR

Before I committed code, I have already rebased the code and updated the perl tidy version.

kalikiana commented on t/config.t:
```diff
@@ -147,7 +147,28 @@ subtest 'Test configuration override from file' => sub {
+    my $t_dir = tempdir;
+    local $ENV{OPENQA_CONFIG} = $t_dir;
+    my $app  = Mojolicious->new();
+    my @data = (
```
I'd find multi-line strings here easier to read ie.

    my $data = `
        [global]
        appname =    openQA   # foo
    `;

And you could include comments as well.

Amrysliu replied:
I will using the multi-line string, but comments are not supported. The code does support the value with "#" comments. Do we need to support it?

okurz replied:
I don't think we need to support in-line comments after the key=value part, only complete commented line, if this is what you mean. This would also correspond to what is usually supported in .ini-format AFAIK

Amrysliu replied:
In this pr, I prefer not support in-line comments after key=value.

okurz commented on lib/OpenQA/Setup.pm:
```diff
@@ -222,10 +223,11 @@ sub read_config {
         }
     }
-    $app->config->{global}->{recognized_referers} = [split(/ /, $app->config->{global}->{recognized_referers})];
+
```
could you please delete this blank line? I think it sneaked in by mistake

Amrysliu replied:
ok, I will delete it

okurz commented on t/api/05-machines.t:
```diff
@@ -163,6 +163,69 @@ is_deeply(
+    $t->post_ok(
+        '/api/v1/machines',
+        form => {
+            name                => " createwithspace ",
```
would you mind making that more readable with underscore as word separators?

Amrysliu replied:
I will modify those names

Martchus commented:
The others gave already feedback. I'd just like to note that I did similar refactoring in https://github.com/os-autoinst/openQA/pull/2074. Likely I will have to resolve some conflicts then.

Amrysliu commented:
> I think you need to rebase after we have updated the code style in another PR

Before I committed code, I have already rebased the code and updated the perl tidy version.

Amrysliu commented:
> The others gave already feedback. I'd just like to note that I did similar refactoring in #2074. Likely I will have to resolve some conflicts then.

I see #2074 , I did not know you have refactored the setting code in adminTable.  That make trouble for you. You will resolve the conflicts or let me do this?

okurz commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -320,4 +282,51 @@ sub destroy {
     $self->render(json => {result => int($ret)});
 }
 
+=over 4
```
Hm, for other POD in the same file and also others we use a different style for method headers. Would you mind using the same style here? E.g. head2, I think

okurz replied:
I just checked, sorry, disregard my comment. The file already uses the same style in another place. leave everything as is :)

kraih replied:
I don't think we actually have a consistent documentation style for openQA. Might make sense to agree on a standard though.

Martchus commented:
I will take care of the conflict resolution myself.

kraih commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -320,4 +282,51 @@ sub destroy {
+
+    for my $par (@{$tables{$table}->{required}}) {
+        $validation->required($par);
+        if (!defined $self->param($par)) {
```
If you're using validation you need to use `$validation->param()` instead of `$self->param()` consistently.

Amrysliu replied:
I will fix it.

kalikiana commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -320,4 +282,51 @@ sub destroy {
+            $error .= " $par" if $validation->has_error($par);
+        }
+        $self->render(json => {error => $error}, status => 400);
+        return;
```
This seems problematic. An error is rendered, not sure what will be returned here, and `create` will probably continue with no settings?

I'd find it cleaner to `die` here, `catch` it along with the database call and only render in the consuming function.

Martchus replied:
If the `return;` is a `return undef;` and the consuming function handles the `undef`-case correctly I'm actually fine with this.

kraih replied:
Then you can't pass the error message along though. But have to agree that this needs an explicit return value the caller can handle appropriately.

Amrysliu replied:
I remove return self-render in this inner function, only return an error message when check parameter failed.

kraih commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -320,4 +284,51 @@ sub destroy {
+    if ($validation->has_error) {
+        $error_message = "wrong parameter:";
+        for my $par (@{$tables{$table}->{required}}) {
+            $error_message .= " $par" if $validation->has_error($par);
```
A comma separated list would have looked better here.

Amrysliu replied:
I will fix it.

kraih commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -320,4 +284,51 @@ sub destroy {
+
+    my $error_message;
+    if ($validation->has_error) {
+        $error_message = "wrong parameter:";
```
Since you seem to only validate presence of a parameter, wouldn't `Missing parameter: ` be more correct and helpful to the user?

kraih commented on t/api/05-machines.t:
```diff
@@ -84,8 +84,9 @@ is_deeply(
+is($t->tx->res->json->{error}, 'wrong parameter: backend', 'miss backend');
 $res = $t->post_ok('/api/v1/machines', form => {backend => "kde/usb"})->status_is(400);        #missing name
-
+is($t->tx->res->json->{error}, 'wrong parameter: name', 'miss name');
```
This would look nicer as `$t->json_is('/error', 'wrong parameter: name')`.

Amrysliu replied:
Modified it.

kraih commented on t/api/05-machines.t:
```diff
@@ -158,6 +159,69 @@ is_deeply(
+                        }]}]
+        },
+        'trim whitespace characters when create table'
+    ) || diag explain $t->tx->res->json;
```
This might also be nicer as `$t->json_is({...})->or(sub { diag ... })`.

Amrysliu replied:
modified

kraih commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -320,4 +284,52 @@ sub destroy {
+            $error_message .= "$par, " if $validation->has_error($par);
+        }
+        $error_message =~ s/,\s$//;
+        return $error_message;
```
This is much more inefficient than it has to be, always use `join` and an array when making comma separated lists.
```perl
    if ($validation->has_error) {
        my @keys = grep { $validation->has_error($par) } @{$tables{$table}->{required}};
        return "Missing parameters: " . join(', ', @keys);
    }
```

Amrysliu replied:
I fixed it

kalikiana commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -320,4 +284,48 @@ sub destroy {
+        $entry->{$par} = trim $validation->param($par);
+    }
+
+    my $error_message;
```
This doesn't seem to be used anymore

Amrysliu replied:
if I remove it, the following return should be modified to `return (undef, \@settings, \@keys)`, is it ok?

kraih commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -320,4 +284,47 @@ sub destroy {
+=item prepare_settings()
+
+Internal method to prepare settings when add or update admin table.
+Use by both B<create()> and B<update()> method.
```
If this is an internal method then the name should be `_prepare_method`. Or is there another reason it's part of the public API?

Amrysliu replied:
yes, this is an internal method. I will change the method name

kraih commented on lib/OpenQA/WebAPI/Controller/API/V1/Table.pm:
```diff
@@ -320,4 +284,47 @@ sub destroy {
+
+    if ($validation->has_error) {
+        my @miss_keys = grep { $validation->has_error($_) } @{$tables{$table}->{required}};
+        return "Missing parameter: " . join(', ', @miss_keys);
```
`return "Missing parameter: " . join(', ', @{$validation->failed});` is more efficient.

Amrysliu replied:
I will modify it

kraih commented on t/api/05-machines.t:
```diff
@@ -84,7 +84,11 @@ is_deeply(
 
 my $res = $t->post_ok('/api/v1/machines', form => {name => "testmachine"})->status_is(400);    # missing backend
+$t->json_is('/error', 'Missing parameter: backend');
 $res = $t->post_ok('/api/v1/machines', form => {backend => "kde/usb"})->status_is(400);        #missing name
```
`$res` seems unused everywhere. And you can make a method chain like `$t->post_ok(...)->status_is(...)->json_is(...)`. That usually looks better and shows that the tests belong together.

Amrysliu replied:
I changed the code to use method chain. About $res, it is a problem left over from history, I think if it is necessary it should be fixed in another PR.

kraih commented on t/api/05-machines.t:
```diff
@@ -83,8 +83,11 @@ is_deeply(
+my $res = $t->post_ok('/api/v1/machines', form => {name => "testmachine"})->status_is(400)
+  ->json_is('/error', 'Missing parameter: backend');    # missing backend
+$t->post_ok('/api/v1/machines', form => {backend => "kde/usb"})->status_is(400)
+  ->json_is('/error', 'Missing parameter: name');       #missing name
```
The comments here don't really add new information, it's already in the error message now.
